keps/sig-node/2000-graceful-node-shutdown/README.md

```diff
@@ -0,0 +1,866 @@
+### Implementation
+
+Introduce a new Kubelet Config setting, `kubeletConfig.ShutdownGracePeriod`,
+defaulting to 0 seconds. Upon kubelet startup,
+
+*   if the setting is greater than 0 seconds
+    *   kubelet will check with dbus current `InhibitDelayMaxSec` to check if
+        `kubeletConfig.ShutdownGracePeriod <=` `InhibitDelayMaxSec`.
+*   if `kubeletConfig.ShutdownGracePeriod` > `InhibitDelayMaxSec`
+    *   Kubelet will attempt to update the InhibitDelayMaxSec setting, by
+        writing a config file to `/etc/systemd/logind.conf.d/kubelet.conf`, and
+        sending a SIGHUP to logind to update the config setting to ensure that
+        the ShutdownGracePeriod from kubelet config is equal to
+        `InhibitDelayMaxSec`.
+
+After updating the `InhibitDelayMaxSec` on the node if needed, Kubelet will
+query the dbus for the final value of `InhibitDelayMaxSec` set on the node and
+treat min(`InhibitDelayMaxSec`, `kubeletConfig.ShutdownGracePeriod`) as the
+allocatable  shutdown grace period, which will be referred to in this KEP as
+`ShutdownGracePeriod`.
+
+Kubelet will register with dbus as a delay systemd inhibitor lock for the
+`ShutdownGracePeriod` for the shutdown event. Kubelet will also register a
+`PrepareForShutdown` signal which will be emitted prior to the shutdown. Upon
+receiving the signal, Kubelet will have additional `ShutdownGracePeriod` time
+before the actual node will initiate the shutdown.
+
+**Handling the shutdown**
+
+Upon a shutdown occurring, Kubelet will gracefully terminate all the pods
+running on the node and update the Ready condition of the node to false with a
+message `Node Shutting Down`, thereby ensuring new workloads will not get
+scheduled to the node.
+
+Since some of the pods running on the node are often critical for the the
+workloads running on a node (e.g. logging pod daemonset, kubeproxy, kubedns)
+etc, we choose to split the pods running on the node into two categories,
+“critical system pods”, and regular pods. Critical system pods should be
+terminated last, because for example, if the logging pod is terminated first,
+logs from the other workloads will not be captured. Critical system pods are
+identified as those that are in the `system-cluster-critical` or
+`system-node-critical` [priority
+classes](https://kubernetes.io/docs/tasks/administer-cluster/guaranteed-scheduling-critical-addon-pods/#marking-pod-as-critical).
+
+Upon shutdown Kubelet will:
+
+1. Gracefully terminate all non critical system pods with a gracePeriodOverride
+computed as `min(podSpec.terminationGracePeriodSeconds, ShutdownGracePeriod)`
+2. Gracefully terminate all critical system pods with gracePeriodOverride of 2
+seconds
+
+Kubelet will use the same existing
+[killPod](https://github.com/kubernetes/kubernetes/blob/release-1.19/pkg/kubelet/pod_workers.go#L292)
+function to perform the termination of pods, using `gracePeriodOverride` to set
+the appropriate grace period. During the termination process, normal [pod
```

Comment:
Hi! Rodrigo here, we discussed the coordination with sidecar KEP. How are you doing? :)

As part of this KEP or sidecar we need to make a change to honor `gracePeriodOverride` ([current code][link] waits for pod.TermiantionGracePeriod to run the pre-stop hooks). Can you do it yourself (should be like 1 line change :))? It is inlikely that the sidecar KEP will be ready for 1.20, I can help if you want, though :)


[link]: https://github.com/kubernetes/kubernetes/blob/1b4fa94c2131dd0b11b470a280b967b0b10ffe2d/pkg/kubelet/kuberuntime/kuberuntime_container.go#L604-L626

Reply:
Good to hear from you Rodrigo, hope you're doing well!

Thanks for in pointing that, out added a note to KEP, looks like https://github.com/kubernetes/kubernetes/issues/92432 is the corresponding issue. I think it makes to change it as part of this KEP, since this it'll impact the behavior here and for sidecar proposal.